mods/carts/cart_entity.lua

```diff
@@ -173,17 +173,17 @@ local function rail_on_step(self, dtime)
 		return
 	end
 
+	local pos = self.object:getpos()
+	local update = {}
+
 	-- stop cart if velocity vector flips
-	if self.old_vel and (((self.old_vel.x * vel.x) < 0) or
-			((self.old_vel.z * vel.z) < 0)) and
-			(self.old_vel.y == 0) then
-		self.old_dir = {x = 0, y = 0, z = 0}
+	if self.old_vel and self.old_vel.y == 0 and
+			(self.old_vel.x * vel.x < 0 or self.old_vel.z * vel.z < 0) then
 		self.old_vel = {x = 0, y = 0, z = 0}
-		self.velocity = {x = 0, y = 0, z = 0}
```

Comment:
Note that this variable is only used once, right after the cart was punched to increase the cart's velocity.